lib/absinthe/type/interface.ex

```diff
@@ -102,7 +102,9 @@ defmodule Absinthe.Type.Interface do
       end
     else
       type_name =
-        Enum.find(implementors, fn type ->
+        implementors
```

Comment:
I think it's actually this list which is wrong. The implementors list should be only concrete implementors. This change should occur in lib/absinthe/blueprint/schema/interface_type_definition.ex `find_implementors/1`

Reply:
This way the filtering is accomplished at schema building time instead of every time the interface is evaluated at runtime.

Reply:
I'm running out of time now, I'll nave a look later 👍

Reply:
@benwilson512 If I change it in `finde_implementors`, I break those two tests:

```
  1) test interfaces are propagated across type imports (Absinthe.Schema.Rule.ObjectMustImplementInterfacesTest)
     test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:88
     Assertion with == failed
     code:  assert %{named: [:cat, :dog, :user], favorite_foods: [:cat, :dog, :user], parented: [:named]} == Schema.__absinthe_interface_implementors__()
     left:  %{favorite_foods: [:cat, :dog, :user], named: [:cat, :dog, :user], parented: [:named]}
     right: %{favorite_foods: [:cat, :dog, :user], named: [:cat, :dog, :user], parented: []}
     stacktrace:
       test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:89: (test)

.

  2) test interfaces are set from sdl (Absinthe.Schema.Rule.ObjectMustImplementInterfacesTest)
     test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:122
     Assertion with == failed
     code:  assert %{image: [], node: [:image, :resource], resource: [:image]} == InterfaceImplementsInterfaces.__absinthe_interface_implementors__()
     left:  %{image: [], node: [:image, :resource], resource: [:image]}
     right: %{image: [], node: [], resource: []}
     stacktrace:
       test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:123: (test)
```

Do we consider those tests correct or wrong?

Reply:
@benwilson512 In which direction should I continue?

Reply:
@benwilson512 I need some guidance to finish this PR.

Reply:
I just had a look at those 2 tests and I believe the schema for those is wrong. According to the [spec](https://spec.graphql.org/draft/#sel-GAHbhBHABmBBoBz3W) and my understanding of the following sentence:

>Transitively implemented interfaces (interfaces implemented by the interface that is being implemented) must also be defined on an implementing type or interface.

I believe that both `dog` & `cat` objects should declare the `parented` interface as well.

Reply:
yeah, I have a fix locally for #1080 and those schemas came up as being invalid. Just need to clean it all up for a PR for that issue

Reply:
@maennchen I have merged @maartenvanvliet [PR](https://github.com/absinthe-graphql/absinthe/issues/1080) which fixed those 2 tests. Could you give it another try when you get the chance and let us know how it goes now ? Thanks in advance.

Reply:
@tlvenn I've rebased my commits and reverted my implementation changes. The tests are failing.

Reply:
@tlvenn 

> I believe the schema for those is wrong

What would you specifically change in the test?

Reply:
The schema used by the test has been fixed so that `dog` `cat` & `user` are declaring the `parented` interface as they should be.

Reply:
And the tests themselves have been updated to assert the right thing as well.

Reply:
If you rebase on master your PR I believe the tests should work now.

Reply:
@tlvenn Can you please specify exactly what you mean I should change in my test? I can't follow what you think should be different.

I did rebase on master. It does not pass.

Reply:
Interesting... Can you update your PR with the rebase and can you copy paste the results of those 2 tests please ?

Reply:
Ha i see that you already pushed the rebase my bad, can you just update in the comment the results of those 2 tests ? Thanks !

Reply:
Actually after you pushed your rebased PR, only the new test that is correctly failing now remains but the fix in `interface_type_definition` is gone or you never actually pushed it.

Reply:
@tlvenn I removed the fix and only left the test, since I thought it should already be fixed on master.

I'll reapply the fix then. :)

Reply:
Sorry for the confusion, btw, It seems you pushed your initial fix and not the one targeting lib/absinthe/blueprint/schema/interface_type_definition.ex `find_implementors/1`

Reply:
@tlvenn Now I finally get it. :smile: 

I'll try that way then

Reply:
@tlvenn 

https://github.com/absinthe-graphql/absinthe/runs/4335029279?check_suite_focus=true

```
  1) test interfaces are set from sdl (Absinthe.Schema.Rule.ObjectMustImplementInterfacesTest)
Error:      test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:132
     Assertion with == failed
     code:  assert %{image: [], node: [:image, :resource], resource: [:image]} == InterfaceImplementsInterfaces.__absinthe_interface_implementors__()
     left:  %{image: [], node: [:image, :resource], resource: [:image]}
     right: %{image: [], node: [], resource: []}
     stacktrace:
       test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:133: (test)

.

  2) test interfaces are propagated across type imports (Absinthe.Schema.Rule.ObjectMustImplementInterfacesTest)
Error:      test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:98
     Assertion with == failed
     code:  assert %{named: [:cat, :dog, :user], favorite_foods: [:cat, :dog, :user], parented: [:cat, :dog, :named, :user]} == Schema.__absinthe_interface_implementors__()
     left:  %{favorite_foods: [:cat, :dog, :user], named: [:cat, :dog, :user], parented: [:cat, :dog, :named, :user]}
     right: %{
              favorite_foods: [:cat, :dog, :user],
              named: [:cat, :dog, :user],
              parented: [:cat, :dog, :user]
            }
     stacktrace:
       test/absinthe/schema/rule/object_must_implement_interfaces_test.exs:99: (test)
```

You seem to know that code better. Should I just give you push access to my fork?

Reply:
Ha great, thanks for this, I wanted to believe it would work but kinda suspected this could happen. Those 2 issues have been introduced by the original PR that added support for interface implementing interfaces because until now `find_implementors/1` was incorrectly returning interfaces instead of just concrete types.

Given that:
- This test suite is purposely for testing objects that use interfaces.
- We have other tests against interfaces implementing interfaces in `test/absinthe/schema/rule/object_interfaces_must_be_valid_test.exs`.
- Transitively implemented interfaces  must also be defined on implementing types.

We should simply get rid of interfaces implementing interfaces in those 2 tests and fix the assertions, we are mixing too many concerns otherwise and the purpose of those tests is lost.

We probably could / should introduce a specific test to assert that `find_implementors/1` is only returning concrete types in another place.

I am totally happy to take it from there if you want, totally up to you and in any case, thanks again for your time and effort on this. Have a great WE !

Reply:
@maennchen can you please re send  the invitation to collab on your fork, i did not see it initially and it has expired now, thanks.

Reply:
@tlvenn I sent it again :)